feat(manager/maven-wrapper): add private packages support

sleicht opened the pull request:
Add support for private Maven repositories when updating maven-wrapper by setting MVNW_USERNAME and MVNW_PASSWORD environment variables from hostRules configuration.

Maven Wrapper natively supports these environment variables for authentication when downloading the Maven distribution. The implementation:

- Collects registry URLs from custom wrapper URL and dependency registryUrls
- Queries hostRules for credentials matching each registry
- Sets MVNW_USERNAME and MVNW_PASSWORD for the first registry with credentials
- Works seamlessly with existing MVNW_REPOURL functionality

<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Downloading maven wrapper will now work as well for custom environments where username/pw is needed.

## Context

Please select one of the below:

- [X] This closes an existing Issue: #25805
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [X] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository



<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

sleicht committed:
feat(manager/maven-wrapper): add private packages support (#25805)

Add support for private Maven repositories when updating maven-wrapper
by setting MVNW_USERNAME and MVNW_PASSWORD environment variables from
hostRules configuration.

Maven Wrapper natively supports these environment variables for
authentication when downloading the Maven distribution. The
implementation:

- Collects registry URLs from custom wrapper URL and dependency
  registryUrls
- Queries hostRules for credentials matching each registry
- Sets MVNW_USERNAME and MVNW_PASSWORD for the first registry with
  credentials
- Works seamlessly with existing MVNW_REPOURL functionality

Churro reviewed (changes requested):
First review of the code itself. Not ready to merge, also in the light that the AI-generated code obviously hasn't been thoroughly scrutinized before committing.

sleicht committed:
Merge branch 'main' into feat/25805-maven-wrapper-private-packages

sleicht commented on lib/modules/manager/maven-wrapper/artifacts.ts:
```diff
@@ -166,11 +168,67 @@ async function executeWrapperCommand(
+      url: registryUrl,
+    });
+
+    if (username || password) {
```
Correct – changed to only set the variables if both are present and log a warning if only one is present.

sleicht committed:
refactor(manager/maven-wrapper): require both username and password for authentication (#25805)

Only set MVNW_USERNAME and MVNW_PASSWORD when both credentials are
present. Replace registryUrls with replaceString in test fixtures and
simplify credential lookup to use customMavenWrapperUrl directly.

sleicht commented:
Thank you @Churro for the review. I apologize for not thinking through the AI changes before creating this PR.

sleicht committed:
refactor(manager/maven-wrapper): remove redundant credential checks (#25805)

Remove unnecessary nested checks for username and password since they are already verified by the parent condition.

Churro commented on lib/modules/manager/maven-wrapper/artifacts.spec.ts:
```diff
@@ -474,4 +478,194 @@ describe('modules/manager/maven-wrapper/artifacts', () => {
+      clearHostRules();
+    });
+
+    it('should not set credentials when only username is configured', async () => {
```
Test superfluous, see below

Churro commented on lib/modules/manager/maven-wrapper/artifacts.ts:
```diff
@@ -166,11 +168,35 @@ async function executeWrapperCommand(
-    return { MVNW_REPOURL: customMavenWrapperUrl };
+    extraEnv.MVNW_REPOURL = customMavenWrapperUrl;
+
+    // Set credentials for Maven wrapper repository
```
```suggestion
```
No need to explain obvious behavior

Churro commented on lib/modules/manager/maven-wrapper/artifacts.ts:
```diff
@@ -166,11 +168,35 @@ async function executeWrapperCommand(
+    if (username && password) {
+      extraEnv.MVNW_USERNAME = username;
+      extraEnv.MVNW_PASSWORD = password;
+      logger.debug(
```
```suggestion
      logger.trace(
```

Churro commented on lib/modules/manager/maven-wrapper/artifacts.ts:
```diff
@@ -9,7 +9,9 @@ import type { ExecOptions, ExtraEnv } from '../../../util/exec/types';
 import { chmodLocalFile, readLocalFile, statLocalFile } from '../../../util/fs';
 import { getRepoStatus } from '../../../util/git';
 import type { StatusResult } from '../../../util/git/types';
+import { find as findHostRules } from '../../../util/host-rules';
```
Nitpick but why `findHostRules`, when in fact only one `HostRule` is returned?

Please just do `import * as hostRules from '../../../util/host-rules';` 
and later on `const { username, password } = hostRules.find({`. 

It's only a minor thing but helps to avoid introducing X different naming and usage patterns for the same function throughout the whole code base. The pattern I mentioned is the one that's already established.

sleicht committed:
refactor(manager/maven-wrapper): simplify tests and improve logging (#25805)

Remove redundant describe block, reduce test duplication, use trace logging for credentials, and simplify host rules import.

Churro reviewed (changes requested):
Otherwise LGTM

sleicht committed:
refactor(manager/maven-wrapper): remove duplicate credential test (#25805)

Remove duplicate test case for custom repository URL credentials as this scenario is already covered by existing tests.

RahulGautamSingh reviewed (changes requested):
Else, LGTM

sleicht committed:
refactor(manager/maven-wrapper): change credentials logging from trace to debug

Update log level when setting MVNW_USERNAME and MVNW_PASSWORD to make credential usage more visible during execution.

sleicht committed:
Merge branch 'main' into feat/25805-maven-wrapper-private-packages

viceice commented on lib/modules/manager/maven-wrapper/artifacts.spec.ts:
```diff
@@ -5,6 +5,10 @@ import upath from 'upath';
 import { GlobalConfig } from '../../../config/global';
 import { resetPrefetchedImages } from '../../../util/exec/docker';
+import {
+  add as addHostRules,
```
```suggestion
  add as addHostRule,
```
adds a single rule

viceice commented on lib/modules/manager/maven-wrapper/artifacts.ts:
```diff
@@ -166,11 +168,28 @@ async function executeWrapperCommand(
+    if (username && password) {
+      extraEnv.MVNW_USERNAME = username;
+      extraEnv.MVNW_PASSWORD = password;
+      logger.trace(
```
why this needs to be debug? it could be remapped if needed 🤔

viceice replied:
why this needs to be debug? it could be remapped if needed 🤔

sleicht replied:
Are you questioning whether this change from "trace" to "debug" is necessary, or if this log is needed at all?

sleicht commented on lib/modules/manager/maven-wrapper/artifacts.spec.ts:
```diff
@@ -490,4 +494,52 @@ describe('modules/manager/maven-wrapper/artifacts', () => {
+    });
+
+    // Verify environment variables are set
+    expect(execSnapshots[0].options!.env).toHaveProperty(
```
Thanks, thats way better.

sleicht committed:
refactor(manager/maven-wrapper): improve test clarity and setup

Move clearHostRules to beforeEach, use toMatchObject for cleaner assertions, and rename import for consistency.

sleicht committed:
Merge branch 'main' into feat/25805-maven-wrapper-private-packages

sleicht committed:
Merge branch 'main' into feat/25805-maven-wrapper-private-packages

* main: (353 commits)
  fix(pep440): handle v-prefixed versions gracefully (#40989)
  feat(presets/workarounds): support bellsoft/hardened-liberica-runtime-container Image (#41058)
  feat(deps): update ghcr.io/renovatebot/base-image docker tag to v13.4.0 (main) (#41079)
  build(deps): update dependency glob to v13.0.1 (main) (#41074)
  chore(deps): update dependency @biomejs/biome to v2.3.14 (main) (#41069)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.3.4 (main) (#41070)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.3.3 (main) (#41067)
  chore(deps): update dependency node to v24.13.1 (main) (#41066)
  build(deps): update node.js to v24.13.1 (main) (#41060)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.3.2 (main) (#41061)
  chore(tools/mkdocs): ensure `RENOVATE_VERSION` is passed to `mkdocs` (#40988)
  fix(versioning/redhat): allow optional .GA inside version (#41059)
  feat(preset): update :semanticPrefixFixDepsChoreOthers preset with missing poetry manager (#41049)
  fix(types): RenovateRepository not inheriting RenovateConfig (#40897)
  fix(types): don't use `infer` from Zod (#41054)
  chore(deps): update dependency @types/node to v24.10.10 (main) (#41050)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.3.1 (main) (#41048)
  fix(vendir): remove registryUrls from git-refs dependencies (#40938)
  chore(deps): update dependency @smithy/util-stream to v4.5.11 (main) (#41046)
  chore(deps): lock file maintenance (main) (#41043)
  ...

sleicht committed:
refactor: Rewrite imports with `.ts` extensions (#40700)

viceice reviewed (commented):
otherwise LGTM

sleicht committed:
refactor(maven-wrapper): relocate StatusResult import to git/types module

sleicht committed:
Merge remote-tracking branch 'origin/main' into feat/25805-maven-wrapper-private-packages

* origin/main: (615 commits)
  chore(deps): update linters to v1.56.0 (main) (#42071)
  ci: remove unused Trivy workflow (#42067)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.27.6 (main) (#42066)
  chore(deps): update ghcr.io/containerbase/devcontainer docker tag to v14.6.7 (main) (#42064)
  docs: update references to renovatebot/github-action to v46.1.6 (main) (#42063)
  chore(deps): update dependency @containerbase/semantic-release-pnpm to v1.3.27 (main) (#42062)
  chore(deps): update dependency @containerbase/istanbul-reports-html to v1.1.37 (main) (#42061)
  chore(deps): update containerbase/internal-tools action to v4.5.4 (main) (#42060)
  docs: update references to python to ffebef4 (main) (#42058)
  docs: update references to otel/opentelemetry-collector-contrib to v0.148.0 (main) (#42059)
  refactor: massage `minimumRelaseAge="0days"` to null (#42053)
  chore(deps): update dependency oxlint-tsgolint to v0.17.0 (main) (#42052)
  docs: add SPDX license ID to home page (#41840)
  feat(presets/monorepos): Add Effect.ts v4 to monorepo.json (#41939)
  chore(deps): update dependency tsdown to v0.21.3 (main) (#42049)
  feat(manager/mise): add prettier to short tool names (#42047)
  chore(deps): update dependency lint-staged to v16.4.0 (main) (#42043)
  chore(deps): update dependency lint-staged to v16.3.4 (main) (#42041)
  build(deps): update dependency better-sqlite3 to v12.8.0 (main) (#42040)
  build(deps): update dependency slugify to v1.6.8 (main) (#42037)
  ...

RahulGautamSingh reviewed (changes requested):
Failing tests

sleicht committed:
fix(manager/maven-wrapper): add -Dtype=script flag to wrapper command in test

sleicht committed:
docs(manager/maven-wrapper): document private registry authentication support

- add section explaining how to configure hostRules for private Maven repositories
- include example configuration with placeholder for encrypted credentials
- link to comprehensive private package support documentation

sleicht committed:
Merge remote-tracking branch 'origin/main' into feat/25805-maven-wrapper-private-packages

* origin/main: (227 commits)
  refactor(vulnerability): check in case-insensitive manner (#42457)
  fix: Revert "build(deps): update dependency global-agent to v4 (main)" (#42458)
  refactor(vulnerability): extract shared functionality into util functions (#42433)
  refactor(lint): Improve pnpm check (#42387)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.33.13 (main) (#42455)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.33.12 (main) (#42454)
  fix(deps): update ghcr.io/renovatebot/base-image docker tag to v13.33.11 (main) (#42453)
  chore(deps): update dependency oxlint-tsgolint to v0.19.0 (main) (#42452)
  feat(presets): Add hocuspocus monorepo (#42426)
  chore(deps): update codecov/codecov-action action to v6 (main) (#42431)
  chore(deps): update actions/deploy-pages action to v5 (main) (#42430)
  chore(lint): add `enforce-ts-extension` oxlint rule (#42389)
  fix(manager/github-actions): disable updates for bare SHA pins without version comment (#42398)
  chore(deps): update dependency vite to v8.0.5 [security] (main) (#42437)
  docs: fix typo (#42427)
  chore(deps): update dependency vite to v8.0.3 (main) (#42429)
  test(util): wrap in a `describe` block (#42432)
  refactor(lint): Switch to Oxlint (#42384)
  test: fix flaky test (#42428)
  fix(git): skip merge commits for semantic commit detection (#42419)
  ...

sleicht commented:
@jamietanna Was the change I made that what you wanted to see in the wrapper doc?

sleicht committed:
Merge branch 'main' into feat/25805-maven-wrapper-private-packages

RahulGautamSingh reviewed (approved):
lgtm
